chore: RHINENG-2212 - Migrate to React 18 & Patternfly v5

bastilian opened the pull request:
This is a follow up to https://github.com/RedHatInsights/compliance-frontend/pull/1991, replacing all remaining tests using enzyme to use react-testing library instead.

It also gets rid of snapshots in favour of properly querying elements and interact with them. 

## Secure Coding Practices Checklist GitHub Link
- https://github.com/RedHatInsights/secure-coding-checklist

## Secure Coding Checklist
- [ ] Input Validation
- [ ] Output Encoding
- [ ] Authentication and Password Management
- [ ] Session Management
- [ ] Access Control
- [ ] Cryptographic Practices
- [ ] Error Handling and Logging
- [ ] Data Protection
- [ ] Communication Security
- [ ] System Configuration
- [ ] Database Security
- [ ] File Management
- [ ] Memory Management
- [ ] General Coding Practices

bastilian commented:
/retest

gkarat commented on .eslintrc.yml:
```diff
@@ -10,13 +10,7 @@ globals:
   describe: true
   expect: true
```
Most probably, we can safely remove the jest global variables from this list since the FEC eslint config already sets "jest" environment to true https://github.com/RedHatInsights/frontend-components/blob/master/packages/eslint-config/index.js#L12

gkarat commented on src/SmartComponents/CreatePolicy/CreatePolicy.test.js:
```diff
@@ -1,42 +1,32 @@
 
 describe('CreatePolicyForm', () => {
   it('expect to render the wizard', () => {
+    // TODO Replace with proper data and extend tests
```
Would be nice to have tests checking the wizard steps as well (the previous tests covered this) so that we don't decrease the coverage.

gkarat commented on src/SmartComponents/CreatePolicy/CreateSCAPPolicy.test.js:
```diff
@@ -1,83 +1,67 @@
+    expect(screen.getByText('Loading...')).toBeInTheDocument();
   });
 
   it('should render benchmarks and no policies until one is selected', () => {
```
This test doesn't check if benchmarks are rendered

gkarat commented on src/SmartComponents/CreatePolicy/FinishedCreatePolicy.test.js:
```diff
@@ -20,23 +28,14 @@ describe('FinishedCreatePolicy', () => {
+      screen.getByRole('heading', { name: 'Creating policy' })
+    ).toBeInTheDocument();
   });
 });
```
Do we want to retain the error state test? Looks like this scenario was removed

gkarat commented on src/SmartComponents/CreatePolicy/ReviewCreatedPolicy.test.js:
```diff
@@ -50,40 +48,16 @@ describe('ReviewCreatedPolicy', () => {
+      screen.getByRole('heading', {
+        name: 'C2S for Red Hat Enterprise Linux 6',
+      })
+    ).toBeInTheDocument();
```
The same question here as to loading and error states

gkarat commented on src/SmartComponents/EditPolicy/EditPolicySystemsTab.test.js:
```diff
@@ -7,18 +11,15 @@ describe('EditPolicySystemsTab', () => {
       osMajorVersion: '7',
       supportedOsVersions: ['1.2', '1.1', '1.3', '1.4'],
     },
-    newRuleTabs: false,
```
Is it deprecated? Should we stop testing this prop?

gkarat commented on src/SmartComponents/EditPolicy/__snapshots__/EditPolicyForm.test.js.snap:
```diff
@@ -102,10 +102,75 @@ exports[`EditPolicyForm expect to render with alter 1`] = `
+        aria-label="Inventory Table"
         bulkselect="[object Object]"
+        columns="function mergedColumns(defaultColumns) {
+    /* istanbul ignore next */
```
Woah, is it expected?

bastilian replied:
Nah, I think this should be cleaned up in a later commit.

gkarat replied:
yeah, I commented on outdated code, sorry :D

gkarat commented on src/SmartComponents/SystemDetails/SystemDetails.test.js:
```diff
@@ -39,34 +44,26 @@ describe('SystemDetails', () => {
 
-    expect(toJson(wrapper)).toMatchSnapshot();
+    expect(screen.getByText('Loading...')).toBeInTheDocument();
   });
```
Can we keep these tests (especially, with the error scenario)? Or you think it's not relevant/necessary?

gkarat commented on src/SmartComponents/Reports/Reports.js:
Please update props for SkeletonTable: they were updated and none of the current props is registered in the component interface

gkarat reviewed (changes requested):
- Can we check `/src/PresentationalComponents/ComplianceScore/ComplianceScore.js`, `/src/PresentationalComponents/UnsupportedSSGVersion/UnsupportedSSGVersion.js`, `/src/SmartComponents/ReportDetails/Components/ChartLegend.js` and probably other modules where we don't wrap icons with PF's `Icon` component: if the icons there keep the expected color and size?
- It looks like `/src/PresentationalComponents/ProfileThresholdField/ProfileThresholdField.js`FormGroup helper text was not changed to utilise FormHelperText. The helperText is deprecated
- In `/src/SmartComponents/CreatePolicy/EditPolicyDetails.js`, `onChange` property for `TextInput` most probably was not updated (the function now received an event object as a first argument for the callback). This results in a bug that I no longer can edit policy details.

gkarat reviewed (commented):
I am not sure if the bug was introduced by the recent migration changes, but the selection box doesn't show selected items number on any of the views in Compliance.

https://github.com/RedHatInsights/compliance-frontend/assets/31385370/f45111ad-0b7a-4aa5-bddc-65703315b2dc

adonispuente commented:
Overall functionality doesnt seem to be affected. Its just some small ux things.

There are very few ux things I wanted to point out for future refinement tickets (these can already we known though).

- The empty state on pf5 version has a much smaller logo (all components in general are a little smaller. This just might be a pf5 css thing though)
- The business objective section in policy details page has 2 bullet points instead of 1

adonispuente commented on src/PresentationalComponents/RulesTable/components/SeverityIcons.js:
```diff
@@ -0,0 +1,48 @@
+import React from 'react';
```
Nice refactor

adonispuente commented on src/PresentationalComponents/index.js:
```diff
@@ -33,10 +33,10 @@ export { default as ReportsEmptyState } from './ReportsEmptyState/ReportsEmptySt
-  default as NoResultsTable,
-  emptyRows,
-} from './NoResultsTable/NoResultsTable';
+// export {
```
why comment this out instead of delete?

bastilian replied:
Removed it.

adonispuente commented on src/SmartComponents/DeletePolicy/DeletePolicy.js:
```diff
@@ -99,7 +99,7 @@ const DeletePolicy = () => {
             id={`deleting-policy-check-${id}`}
             isChecked={deleteEnabled}
-            onChange={setDeleteEnabled}
+            onChange={(_event, val) => setDeleteEnabled(val)}
```
also small nitpick, better for this to be consistent with your other onChange functions `{(_,v)}`

bastilian replied:
Changed it.

adonispuente reviewed (approved):
Left a few super small nitpicks, when testing against master I couldnt find and functionality broken changes. Just small ux stuff that I believe we're already aware of. Great job overall, LGTM!

bastilian commented:
@gkarat:

> Can we check /src/PresentationalComponents/ComplianceScore/ComplianceScore.js, /src/PresentationalComponents/UnsupportedSSGVersion/UnsupportedSSGVersion.js, /src/SmartComponents/ReportDetails/Components/ChartLegend.js and probably other modules where we don't wrap icons with PF's Icon component: if the icons there keep the expected color and size?

I changed the icons and wrapped them in `Icon` components and made sure they still retain the same (or very close) colour and size. In some cases for example the empty state there seem to be issues still and we need to follow up on those.

> It looks like /src/PresentationalComponents/ProfileThresholdField/ProfileThresholdField.jsFormGroup helper text was not changed to utilise FormHelperText. The helperText is deprecated

Updated to use the PF5 HelperText components. 

> In /src/SmartComponents/CreatePolicy/EditPolicyDetails.js, onChange property for TextInput most probably was not updated (the function now received an event object as a first argument for the callback). This results in a bug that I no longer can edit policy details.

Changed the `onChange`s. All Form components should work as expected. 

Re skipped cypress tests: I left them still skipped and will follow up on those later.

adonispuente reviewed (approved):
All of my comments have been addressed. Just Georgii's comments now,
LGTM!

gkarat commented on src/SmartComponents/EditPolicyDetails/EditPolicyDetailsInline.js:
It feels like something is odd with EditPolicyDetailsInline.js: if you navigate to policy details, you'd see the text for each field duplicated.

gkarat reviewed (approved):
LGTM! Thanks for this huge work @bastilian. There are some open items, but almost all of them we can address in the next PRs. I am not sure I can spot anything else at the moment, the app works just fine after the update.
